src/groupeddataframe/grouping.jl

```diff
@@ -115,71 +118,102 @@ Base.getindex(gd::GroupedDataFrame, I::AbstractArray{Bool}) =
 Base.names(gd::GroupedDataFrame) = names(gd.parent)
 _names(gd::GroupedDataFrame) = _names(gd.parent)
 
-##############################################################################
-##
-## GroupApplied...
-##    the result of a split-apply operation
-##    TODOs:
-##      - better name?
-##      - ref
-##      - keys, vals
-##      - length
-##      - start, next, done -- should this return (k,v) or just v?
-##      - make it a real associative type? Is there a need to look up key columns?
-##
-##############################################################################
-
-"""
-The result of a `map` operation on a GroupedDataFrame; mainly for use
-with `combine`
-
-Not meant to be constructed directly, see `groupby` abnd
-`combine`. Minimal support is provided for this type. `map` is
-provided for a GroupApplied object.
-
-"""
-struct GroupApplied{T<:AbstractDataFrame}
-    gd::GroupedDataFrame
-    vals::Vector{T}
-
-    function (::Type{GroupApplied})(gd::GroupedDataFrame, vals::Vector)
-        length(gd) == length(vals) ||
-            throw(DimensionMismatch("GroupApplied requires keys and vals be of equal length (got $(length(gd)) and $(length(vals)))."))
-        new{eltype(vals)}(gd, vals)
-    end
-end
-
-
-#
-# Apply / map
-#
-
-# map() sweeps along groups
-function Base.map(f::Function, gd::GroupedDataFrame)
-    GroupApplied(gd, [wrap(f(df)) for df in gd])
-end
-function Base.map(f::Function, ga::GroupApplied)
-    GroupApplied(ga.gd, [wrap(f(df)) for df in ga.vals])
-end
-
 wrap(df::AbstractDataFrame) = df
+wrap(nt::NamedTuple) = nt
 wrap(A::Matrix) = convert(DataFrame, A)
-wrap(s::Any) = DataFrame(x1 = s)
+wrap(s::Union{AbstractVector, Tuple}) = DataFrame(x1 = s)
+wrap(s::Any) = (x1 = s,)
 
 """
-Combine a GroupApplied object (rudimentary)
+    map(f::Function, gd::GroupedDataFrame)
+
+Apply a function to each group of rows and return a `GroupedDataFrame`.
+
+For each group in `gd`, `f` is passed a `SubDataFrame` view holding the corresponding rows.
+`f` can return a single value, a named tuple, a vector, or a data frame.
+This determines the shape of the resulting data frame:
+- A single value gives a data frame with a single column and one row per group.
+- A named tuple gives a data frame with one column for each field in the named tuple
+  and one row per group.
+- A vector gives a data frame with a single column and as many rows
```

Comment:
The same question for a vector of named tuples as:

```
julia> DataFrame([(a=1,b=2), (a=3,b=4)])
2×2 DataFrame
│ Row │ a     │ b     │
│     │ Int64 │ Int64 │
├─────┼───────┼───────┤
│ 1   │ 1     │ 2     │
│ 2   │ 3     │ 4     │
```

In general - we do not have to handle this but I want to be clear what we want.

Also what happens in case a matrix is passed?

Reply:
Behavior is unchanged by the PR:
```julia
julia> by(d -> [(a=1,), (a=2,)], DataFrame(x=[1, 1, 2, 3]), :x)
6×2 DataFrame
│ Row │ x     │ x1        │
│     │ Int64 │ NamedTup… │
├─────┼───────┼───────────┤
│ 1   │ 1     │ (a = 1,)  │
│ 2   │ 1     │ (a = 2,)  │
│ 3   │ 2     │ (a = 1,)  │
│ 4   │ 2     │ (a = 2,)  │
│ 5   │ 3     │ (a = 1,)  │
│ 6   │ 3     │ (a = 2,)  │
```

If a matrix is returned, it's passed to the `DataFrame` constructor, which creates `x1`... `xn` columns. I'll document this since we have a special method for it.